File: frontend/src/utils/formatters.ts
/**
 * Formatting utilities for dates, numbers, and text.
 */

/**
 * Format a date according to specified format string.
 *
 * @param date - Date to format
 * @param format - Format string ('short', 'long', 'iso', or custom)
 * @returns Formatted date string
 */
export function formatDate(date: Date | string, format: string = 'short'): string {
  const d = typeof date === 'string' ? new Date(date) : date;

  if (isNaN(d.getTime())) {
    return 'Invalid Date';
  }

  switch (format) {
    case 'short':
      // MM/DD/YYYY
      return d.toLocaleDateString('en-US');

    case 'long':
      // Month DD, YYYY
      return d.toLocaleDateString('en-US', {
        year: 'numeric',
        month: 'long',
        day: 'numeric',
      });

    case 'iso':
      // YYYY-MM-DD
      return d.toISOString().split('T')[0];

    case 'time':
      // HH:MM AM/PM
      return d.toLocaleTimeString('en-US', {
        hour: 'numeric',
        minute: '2-digit',
      });

    case 'datetime':
      // MM/DD/YYYY HH:MM AM/PM
      return d.toLocaleString('en-US');

    default:
      return d.toLocaleDateString('en-US');
  }
}

/**
 * Format duration in minutes to human-readable string.
 *
 * @param minutes - Duration in minutes
 * @returns Formatted duration (e.g., "2h 30m", "45m")
 */
export function formatDuration(minutes: number): string {
  if (minutes < 0) {
    return '0m';
  }

  const hours = Math.floor(minutes / 60);
  const mins = minutes % 60;

  if (hours === 0) {
    return `${mins}m`;
  }

  if (mins === 0) {
    return `${hours}h`;
  }

  return `${hours}h ${mins}m`;
}

/**
 * Format a number as a percentage.
 *
 * @param value - Value to format (0-1 for fraction, 0-100 for percentage)
 * @param decimals - Number of decimal places (default: 1)
 * @param isFraction - Whether value is a fraction (0-1) or percentage (0-100)
 * @returns Formatted percentage string (e.g., "75.5%")
 */
export function formatPercentage(
  value: number,
  decimals: number = 1,
  isFraction: boolean = true
): string {
  const percentage = isFraction ? value * 100 : value;
  return `${percentage.toFixed(decimals)}%`;
}

/**
 * Truncate text to maximum length with ellipsis.
 *
 * @param text - Text to truncate
 * @param maxLength - Maximum length (including ellipsis)
 * @param suffix - Suffix to add when truncating (default: "...")
 * @returns Truncated text
 */
export function truncateText(
  text: string,
  maxLength: number,
  suffix: string = '...'
): string {
  if (text.length <= maxLength) {
    return text;
  }

  const truncateAt = maxLength - suffix.length;
  if (truncateAt <= 0) {
    return suffix.slice(0, maxLength);
  }

  return text.slice(0, truncateAt) + suffix;
}

/**
 * Pluralize a word based on count.
 *
 * @param count - Count to determine pluralization
 * @param singular - Singular form of the word
 * @param plural - Plural form (optional, defaults to singular + "s")
 * @returns Pluralized string with count (e.g., "1 item", "2 items")
 */
export function pluralize(
  count: number,
  singular: string,
  plural?: string
): string {
  const word = count === 1 ? singular : (plural || `${singular}s`);
  return `${count} ${word}`;
}

File: backend/cli/utils/formatters.py
"""
Data formatting utilities for CLI output.
"""

from datetime import datetime, date
from typing import Any


def format_datetime(dt: datetime) -> str:
    """Format datetime for display."""
    if dt is None:
        return "-"
    return dt.strftime("%Y-%m-%d %H:%M:%S")


def format_date(d: date) -> str:
    """Format date for display."""
    if d is None:
        return "-"
    return d.strftime("%Y-%m-%d")


def format_bool(value: bool) -> str:
    """Format boolean for display."""
    return "✓" if value else "✗"


def format_percentage(value: float) -> str:
    """Format percentage for display."""
    if value is None:
        return "-"
    return f"{value:.1f}%"


def format_hours(hours: float) -> str:
    """Format hours for display."""
    if hours is None:
        return "-"
    return f"{hours:.1f}h"


def truncate(text: str, max_length: int = 50) -> str:
    """Truncate text to max length."""
    if text is None:
        return "-"
    if len(text) <= max_length:
        return text
    return text[: max_length - 3] + "..."


def format_list(items: list, max_items: int = 5) -> str:
    """Format list for display."""
    if not items:
        return "-"
    if len(items) <= max_items:
        return ", ".join(str(item) for item in items)
    shown = ", ".join(str(item) for item in items[:max_items])
    return f"{shown}, ... (+{len(items) - max_items} more)"
